### DISCUSSION TITLE
How correctly write unity type for reusable components in react hook form ?

### DISCUSSION THREAD

**Discussion by ghost** [react-hook-form/react-hook-form#11343] [Q&A]
I'm encountering issues with TypeScript types in my React component.

I have the following types defined:

`export type UnitFormInputs = PromotionFormInputs | CategoriesFormInputs;

export type PromotionFormInputs = {
    name: string,
    shortDescription: string,
    longDescription: string,
    categories: {
        id: string,
        name: string
    }[],
    symbolCode: string,
    dealer: string,
    offer: string,
    textUnderPromotions: string,
    promocodes: {
        id: string,
        name: string
    }[],
    startDate: string,
    endDate: string,
    regions: {
        id: string,
        name: string
    }[],
    type: string,
    gradientTopColor: string,
    gradientBottomColor: string,
    isActive: boolean,
    isPassedModeration: boolean,
    sort: number,
    detailedImage: ImageData | File,
    previewInFreshPromotionsImage: ImageData | File,
    isShowInMainBanner: boolean,
    freshPromotionsImage: ImageData | File,
    isShowInFreshPromotions: boolean
}

export type CategoriesFormInputs = {
    name: string,
    isActive: boolean,
    image: ImageData | File
}`

And I have a React component as follows:

`interface Props {
    watch: UseFormWatch<UnitFormInputs>,
    inputName: Partial<keyof UnitFormInputs>,
    maxLength?: number,
    register: UseFormRegister<UnitFormInputs>,
    labelName: string,
    type: "input" | "textarea"
    textAreaRows?: number,
    isRequired: boolean,
    getFieldState: UseFormGetFieldState<UnitFormInputs>
}

const FormInput: FC<Props> = ({
                                  watch,
                                  inputName,
                                  maxLength,
                                  register,
                                  labelName,
                                  type,
                                  textAreaRows,
                                  isRequired,
                                  getFieldState
                              }) => {

    return (
        <div className={styles.field}>
            <p className={styles.fieldName}>{labelName} {
                !!maxLength && <span>{watch(inputName, '').toString().length}/{maxLength}</span>}</p>
            {type === 'input' && <input maxLength={maxLength} type="text" placeholder={labelName}
                                        className={`${global.input} ${styles.input} ${getFieldState(inputName).invalid && styles.error} ${getFieldState(inputName).isDirty && styles.valid}`} {...register(inputName, {
                minLength: 1,
                maxLength: maxLength,
                required: isRequired
            })} />}

            {type === 'textarea' && <textarea rows={textAreaRows} maxLength={maxLength} placeholder={labelName}
                                              className={`${global.input} ${styles.input} ${getFieldState(inputName).invalid && styles.error}`}
                                              {...register(inputName, {
                                                  minLength: 1,
                                                  maxLength: maxLength,
                                                  required: isRequired
                                              })} />}
        </div>
    );
};`

When I pass these props from the parent component, I encounter type errors on watch, register, and inputName props. The error message is: "Type 'UseFormWatch' is not assignable to type 'UseFormWatch' FormInput.tsx(12, 5): The expected type comes from property 'watch' which is declared here on type 'IntrinsicAttributes & Props'.

I've defined the useForm hook like this: const { register, handleSubmit, setValue, watch, control, getFieldState, formState: {errors}, } = useForm()

Is there a way to resolve these type errors, or should I create a separate component for my second type?

I tried to use generic but after that I received x10 more errors

**Comment by Moshyfawn** [comment]
This might give some insight on the matter
https://github.com/orgs/react-hook-form/discussions/7433#discussioncomment-2363962
